clang/lib/CodeGen/HLSLBufferLayoutBuilder.cpp

```diff
@@ -95,187 +46,98 @@ HLSLBufferLayoutBuilder::createLayoutType(const RecordType *RT,
     RecordDecls.push_back(D->bases_begin()->getType()->castAsCXXRecordDecl());
   }
 
-  unsigned FieldOffset;
-  llvm::Type *FieldType;
-
-  while (!RecordDecls.empty()) {
-    const CXXRecordDecl *RD = RecordDecls.pop_back_val();
-
-    for (const auto *FD : RD->fields()) {
-      // No PackOffset info at all, or have a valid packoffset/register(c#)
-      // annotations value -> layout the field.
-      const uint32_t PO = OffsetInfo[Index++];
-      if (PO != CGHLSLOffsetInfo::Unspecified) {
-        if (!layoutField(FD, EndOffset, FieldOffset, FieldType, PO))
-          return nullptr;
-        Layout.push_back(FieldOffset);
-        LayoutElements.push_back(FieldType);
-        continue;
-      }
-      // Have PackOffset info, but there is no packoffset/register(cX)
-      // annotation on this field. Delay the layout until after all of the
-      // other elements with packoffsets/register(cX) are processed.
-      DelayLayoutFields.emplace_back(FD, LayoutElements.size());
-      // reserve space for this field in the layout vector and elements list
-      Layout.push_back(UINT_MAX);
-      LayoutElements.push_back(nullptr);
+  SmallVector<std::pair<const FieldDecl *, uint32_t>> FieldsWithOffset;
+  unsigned OffsetIdx = 0;
+  for (const CXXRecordDecl *RD : llvm::reverse(RecordDecls))
+    for (const auto *FD : RD->fields())
+      FieldsWithOffset.emplace_back(FD, OffsetInfo[OffsetIdx++]);
+
+  llvm::stable_sort(FieldsWithOffset, [](const auto &LHS, const auto &RHS) {
+    return CGHLSLOffsetInfo::compareOffsets(LHS.second, RHS.second);
+  });
```

Comment:
Just as in one of my comments above, I wonder if we could save some of the work that is not necessary in the common case:
```suggestion
  if (!OffsetInfo.empty())
    llvm::stable_sort(FieldsWithOffset, [](const auto &LHS, const auto &RHS) {
      return CGHLSLOffsetInfo::compareOffsets(LHS.second, RHS.second);
    });
```